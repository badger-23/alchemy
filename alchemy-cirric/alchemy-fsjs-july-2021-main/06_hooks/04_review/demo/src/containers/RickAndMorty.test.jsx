import React from 'react';
import { render, screen } from '@testing-library/react';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import RickAndMorty from './RickAndMortyList';
import { MemoryRouter } from 'react-router';
import RickAndMortyDetail from './RickAndMortyDetail';

const getCharacterRoute = rest.get(
  'https://rickandmortyapi.com/api/character/:id',
  (req, res, ctx) => {
    return res(
      ctx.json({
        id: 1,
        name: 'Rick Sanchez',
        status: 'Alive',
        species: 'Human',
        origin: {
          name: 'Earth (C-137)',
        },
        location: {
          name: 'Earth (Replacement Dimension)',
        },
        image: 'https://rickandmortyapi.com/api/character/avatar/1.jpeg',
      })
    );
  }
);

const getCharactersRoute = rest.get(
  'https://rickandmortyapi.com/api/character/',
  (req, res, ctx) => {
    return res(
      ctx.json({
        info: { pages: 34 },
        results: [
          {
            id: 1,
            name: 'Rick Sanchez',
            status: 'Alive',
            species: 'Human',
            origin: {
              name: 'Earth (C-137)',
            },
            location: {
              name: 'Earth (Replacement Dimension)',
            },
            image: 'https://rickandmortyapi.com/api/character/avatar/1.jpeg',
          },
        ],
      })
    );
  }
);

const mockServer = setupServer(getCharacterRoute, getCharactersRoute);

describe('RickAndMorty Containers', () => {
  beforeAll(() => mockServer.listen());
  afterAll(() => mockServer.close());

  describe('RickAndMortyList', () => {
    it('renders a list of characters on the page', async () => {
      const { container } = render(
        <MemoryRouter>
          <RickAndMorty />
        </MemoryRouter>
      );

      screen.getByAltText('Loading Rick and Mortys...');

      const ul = await screen.findByRole('list', { name: 'characters' });
      expect(ul).not.toBeEmptyDOMElement();

      // Above is similar to:
      // return waitFor(() => {
      //   const ul = screen.getByRole('list', { name: 'characters' });
      //   expect(ul).not.toBeEmptyDOMElement();
      // });

      expect(container).toMatchSnapshot();
    });
  });

  describe('RickAndMortyDetail', () => {
    it('renders a single character on the page', async () => {
      render(
        <MemoryRouter initialEntries={['/1']}>
          <RickAndMortyDetail />
        </MemoryRouter>
      );

      screen.getByAltText('Loading Rick and Mortys...');

      const character = await screen.findByText('Rick Sanchez');
      expect(character).not.toBeEmptyDOMElement();

      const characterCard = await screen.findByRole('figure', {
        name: 'character',
      });
      expect(characterCard).toMatchSnapshot();
    });
  });
});
