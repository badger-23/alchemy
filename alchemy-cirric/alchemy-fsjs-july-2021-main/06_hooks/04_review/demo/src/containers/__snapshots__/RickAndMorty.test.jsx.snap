// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RickAndMorty Containers RickAndMortyDetail renders a single character on the page 1`] = `
<figure
  aria-label="character"
  class="character"
>
  <img
    alt="Rick Sanchez"
    class="characterImage"
    src="https://rickandmortyapi.com/api/character/avatar/1.jpeg"
  />
  <figcaption>
    <h3>
      Rick Sanchez
    </h3>
    <p>
      Human
    </p>
    <p>
      Status: 
      Alive
    </p>
    <p>
      Origin: 
      Earth (C-137)
    </p>
    <p>
      Current Location: 
      Earth (Replacement Dimension)
    </p>
  </figcaption>
</figure>
`;

exports[`RickAndMorty Containers RickAndMortyList renders a list of characters on the page 1`] = `
<div>
  <button
    aria-label="Go to previous page"
    disabled=""
  >
    &lt;
  </button>
   
  1
   / 
  34
   
  <button
    aria-label="Go to next page"
  >
    &gt;
  </button>
  <ul
    aria-label="characters"
  >
    <li>
      <a
        href="/1"
      >
        <figure
          class="character"
        >
          <img
            alt="Rick Sanchez"
            class="characterImage"
            src="https://rickandmortyapi.com/api/character/avatar/1.jpeg"
          />
          <figcaption>
            <h3>
              Rick Sanchez
            </h3>
            <p>
              Human
            </p>
            <p>
              Status: 
              Alive
            </p>
          </figcaption>
        </figure>
      </a>
    </li>
  </ul>
</div>
`;
